sumpy/tools.py

```diff
@@ -976,19 +981,198 @@ def wait(self):
+                    + exp * temp[ifft + nfft * (iN2*N1 + iN1_sum)]),
+                id=f"update_{ilev}",
+                depends_on=frozenset([f"exp_{ilev}"])),
+        ]
+
+        domains += [
+            f"[ifft_{ilev}]: 0<=ifft_{ilev}<{nfft}",
+            f"[iN1_{ilev}]: 0<=iN1_{ilev}<{N1}",
+            f"[iN1_sum_{ilev}]: 0<=iN1_sum_{ilev}<{N1}",
+            f"[iN2_{ilev}]: 0<=iN2_{ilev}<{N2}",
+            f"[i_{ilev}]: 0<=i_{ilev}<{n}",
+            f"[i2_{ilev}]: 0<=i2_{ilev}<{n}",
+        ]
+
+    for idom, dom in enumerate(domains):
+        if not dom.startswith("{"):
+            domains[idom] = "{" + dom + "}"
+
+    if n == 1:
+        insns += ["x[:, i] = x[:, i] {dup=i}"]
+    elif inverse:
+        insns += [
+            lp.Assignment(
+                assignee=x[(*broadcast_dims, i3)],
+                expression=x[(*broadcast_dims, i3)] / n,
+                depends_on=frozenset([f"update_{len(factors) - 1}"]),
+            ),
+        ]
+
+    if name is None:
+        if inverse:
+            name = f"ifft_{n}"
+        else:
+            name = f"fft_{n}"
+
+    knl = lp.make_kernel(domains, insns,
+        kernel_data=[
+            lp.GlobalArg("x", shape=shape, is_input=False, is_output=True,
+                dtype=complex_dtype),
+            lp.GlobalArg("y", shape=shape, is_input=True, is_output=False,
+                dtype=complex_dtype),
+            lp.TemporaryVariable("exp_table", shape=(n,),
+                dtype=complex_dtype),
+            lp.TemporaryVariable("temp", shape=(n,),
+                dtype=complex_dtype),
+            ...],
+        name=name,
+        fixed_parameters=fixed_parameters,
+        lang_version=lp.MOST_RECENT_LANGUAGE_VERSION,
+        index_dtype=index_dtype,
+    )
+    if broadcast_dims:
+        knl = lp.split_iname(knl, "j0", 32, inner_tag="l.0", outer_tag="g.0")
+        knl = lp.add_inames_for_unused_hw_axes(knl)
+
+    return knl
+
+
+def _get_fft_backend(queue):
+    env_val = os.environ.get("SUMPY_FFT_BACKEND", None)
+    if env_val:
+        if env_val not in ["loopy", "vkfft"]:
+            raise ValueError("Expected 'loopy' or 'vkfft' for SUMPY_FFT_BACKEND. "
+                   f"Found {env_val}.")
+        return env_val
+
+    try:
+        import pyvkfft.opencl  # noqa: F401
+    except ImportError:
+        warnings.warn("VkFFT not found. FFT runs will be slower.")
+        return "loopy"
+
     if queue.properties & cl.command_queue_properties.OUT_OF_ORDER_EXEC_MODE_ENABLE:
-        raise RuntimeError("VkFFT does not support out of order queues yet.")
+        warnings.warn("VkFFT does not support out of order queues yet. "
+            "Falling back to slower implementation.")
+        return "loopy"
+
+    if (sys.platform == "darwin"
+            and queue.context.devices[0].platform.name
+            == "Portable Computing Language"):
+        warnings.warn("Pocl miscompiles some VkFFT kernels. "
```

Comment:
I would prefer if we had (integration test) coverage of the loopy path even in the case when vkfft works, so that a single `python -m pytest` gives an indication that neither code path is broken.

Reply:
Done in 358318307dc8d8fa245e1e7d14e88cabc3865f2b